ports/mimxrt: Add CYW43 WiFi/BT support.

iabdalkader opened the pull request:
This is a work in progress, attempt at adding CYW43 WiFi/BT support to `ports/mimxrt`, starting with WiFi first and Bluetooth will follow when this is working.

iabdalkader committed:
mimxrt/machine_pin: Extend pin configuration functions.

Add portable pin config macros to mphalport.h.  And add a function to
configure pins with more pin options such as alt function, pull, speed,
drive, interrupt mode, etc.

Note: this new `machine_pin_config()` function can replace
`machine_pin_set_mode()`, but the latter is left as-is to avoid breaking
anything.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
mimxrt: Integrate support for WiFi via the CYW43 driver.

This commit adds the necessary configuration and hooks to get the CYW43
driver working with the mimxrt port.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
mimxrt/sdio: Add SDIO driver.

This is a basic SDIO driver for the mimxrt port, that was added mainly to
support the CYW43 WiFi driver, and as such it only supports the commands
required by the CYW43 driver (but more commands can be added easily). The
driver performs non-blocking DMA transfers, and can detect and recover from
errors.

Note: because the mimxrt port is missing static alternate functions, named
pins and other pin related functions, currently the alternate functions for
USDHC 1 and 2 are hard-coded in the driver.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
mimxrt: Integrate Bluetooth support with NimBLE bindings.

This commit adds the necessary functions to get NimBLE working with the
mimxrt port.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge commented:
> Unrelated question, why is the ticks management in this port so complex ? Why use a GPT timer at all to manage ticks, delay etc.. and not just use the systick like in `stm32` ?

That's a question for @robert-hh

robert-hh commented:
> That's a question for @robert-hh

I had seen the question but thought it was about the STM32 port. If we are talking about ticks.c: that existed when Philipp and me started with the MIMXRT port, and I kept is as it is. But it is a good timing source for µs resolution. It is well made (by @dpgeorge, as git tells), not overly complicated. The calls to the NXP lib look a little bit bulky.

iabdalkader commented:
Yes I was referring to `ticks.c`, I don't think it's needed at all, it uses an extra resource (GPT timer) to handle something that can be handled by the existing and running SysTick, so more code and more power, and it also assumes the crystal is 24MHz, and disables global IRQs in more than one place, and makes a lot of calls to the SDK. IMHO I think it should just be replaced by something like `stm32/systick.c`, or that moved into /shared and reused.

iabdalkader commented:
> But it's working, and as such it follows the rule "if it ain't broken, don't fix it".

I was not going to try to fix anything, just pointing out that there's room for improvement, I'm mainly concerned with WiFi/BT support right now, but you know that rule doesn't really apply here, code changes all the time to improve time/space complexity, even if it's to save a few bytes.

iabdalkader committed:
drivers/cyw43: Make the CYW43 Bluetooth HCI driver more portable.

This commit allows other ports to reuse the CYW43 HCI driver, by replacing
all Bluetooth UART and control named pins with defines in config files and
using `mpbthci` abstract functions (i.e. `mp_bluetooth_hci_*`) instead of
the STM32 specific UART functions.

Note: the function `cywbt_wait_cts_low` does not need to switch the CTS
pin from alternate function to GPIO to read it.  At least on stm32, mimxrt
it's possible to just read the pin input.  For example, see the STM32F7
RM0410 section 6.3.11, and the `SION` for IMXRT.  So this function can
also be available for other ports if the pin mode switching is removed.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
@dpgeorge I think the bluetooth issue might actually be faulty hardware or a power supply issue with the evk/breakout I'm using, I will get prototypes of the actual hardware next week and will retest Bluetooth. In the meantime, I think `cywbt.c` can be made more portable by using the `mp_bluetooth_hci_uart_*` functions instead of the stm32-uart, like I did in the Nina HCI driver, and by defining the Bluetooth control pins in `cyw43_configport.h`.

Regarding the OOB issue, it was actually one of those reads failing and locking up the bus, I've now added code to clear the error after failure, and it works normally now just like the SDIO interrupt mode. Just a thought, it might be valuable to write a porting guide for the driver at some point.

```c
static void cyw43_kso_set(cyw43_int_t *self, int value) {
    // these can fail and it's still ok
    cyw43_write_reg_u8(self, BACKPLANE_FUNCTION, SDIO_SLEEP_CSR, write_value);
    cyw43_write_reg_u8(self, BACKPLANE_FUNCTION, SDIO_SLEEP_CSR, write_value);
``

iabdalkader commented:
@dpgeorge Do we have to switch the CTS pin to GPIO in `cywbt_wait_cts_low` function ? Please see the note in my last commit in this PR for context, alternatively can we move `cywbt_wait_cts_low` to `<port>/mpbthciport.c` with a prototype in `extmod/mpbthci.h` (for example `bool mp_bluetooth_hci_uart_poll_cts(void);`) ?

iabdalkader committed:
mimxrt/irq: Move all IRQ related definitions to dedicated header.

Following other ports, IRQ priorities and related functions are moved to
their own header, to simplify mpconfigport.h.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
@robert-hh I noticed that the processor doesn't wake up from `WFI()` called from `CYW43_DO_IOCTL_WAIT` (see trace below) without SysTick priority set to a lower priority other than the default (lowest),  for some reason GPT IRQ in `ticks.c` doesn't interrupt `WFI()`, this is probably one more reason why it shouldn't be the default timer. Note I also moved the priorities and irq functions to `irq.h` in the last commit to clean up `mpconfigport.h`

```c
cyw43_do_ioctl (self=self@entry=0x2000779c <cyw43_state>, kind=kind@entry=2, cmd=cmd@entry=263, len=len@entry=1560, 
    buf=buf@entry=0x200077d4 <cyw43_state+56> "clmload", iface=iface@entry=0) at ../../lib/cyw43-driver/src/cyw43_ll.c:1223
1223	        CYW43_DO_IOCTL_WAIT;
```

robert-hh commented:
You may have found the reason for a long standing issue in the mimxrt port, in that WFI() was not interrupted by systick. That may have been one reason why ticks.c was implemented this way with the compare IRQ on the GPT. The other reason is µs-precision for delays. Just the ticks_ms_upper counter & handling is a little bit of over-engineering.
Moving priorities and definitions to irq.h is fine, tidying up mpconfigport.h a little bit.
So it seems that what started as "just" a CYW43 port gets a rework for both STM32 and MIMXRT port.

iabdalkader commented:
>  So it seems that what started as "just" a CYW43 port gets a rework for both STM32 and MIMXRT port.

Yeah but I had to get a few things in order first, mainly `cywbt.c`, but hopefully I'm done for this PR.

robert-hh commented:
b.t.w. Which board do you use for testing? I looked back into the various PR about the WFI()/WFE() topic, and it seems only remained for the 117x MCU.

iabdalkader commented:
> b.t.w. Which board do you use for testing? I looked back into the various PR about the WFI()/WFE() topic, and it seems only remained for the 117x MCU.

MIMXRT1060-EVKB

iabdalkader commented:
Final update, WiFi and Bluetooth are both working fine now, see test results below, although for BT nothing above 115200 is reliable, since there's no hardware flow control support in `machine_uart.c`. Fixing that I think is beyond the scope of this PR, as much as I would like to refactor the whole port I have other things to do ;)

Bluetooth
```
./run-multitests.py -i pyb:a0 -i pyb:a1 multi_bluetooth/*py
18 tests performed
16 tests passed
1 tests skipped: multi_bluetooth/ble_deepsleep.py
1 tests failed: multi_bluetooth/stress_log_filesystem.py
```

WiFi
```
./run-multitests.py -i pyb:a0 -i pyb:a1 multi_net/*
20 tests performed
14 tests passed
2 tests skipped: multi_net/tcp_client_rst.py multi_net/uasyncio_tcp_client_rst.py
4 tests failed: multi_net/ssl_cert_rsa.py multi_net/uasyncio_tcp_close_write.py multi_net/uasyncio_tcp_readall.py multi_net/udp_data.py
```

robert-hh commented:
That sounds good, and excellent work. I assume you also verified that BT still works at the PYBD devices.

> Final update,

Never say final, unless you know a word for what comes after final.

>  for BT nothing above 115200 is reliable, since there's no hardware flow control support in machine_uart.c

Did you check that the CYW43 starts to use the flow control signals like CTS at higher baud rates? For the i.MX side that should not be a problem. One can always make rxbuf sufficiently large to hold any possible message. I do not know about the CYW43 part. Implementing hardware flow control should only need some configuration  glue. The NXP drivers support it. I did not add it when creating machine_uart.c, because the respective pins were hardly available at the supported boards.

There is quite a bit at the MIMXRT port than could be reworked, especially the GPIO/Pin handling and the way, the interface configuration is done in the mpconfigboard.h files. These look ugly. Rework of the clock configuration is on my t.b.d. list since ever.

iabdalkader commented:
> I assume you also verified that BT still works at the PYBD devices.

Yes I used it all the time for testing, but also my changes shouldn't have any effect on other ports.

> Never say final, unless you know a word for what comes after final.

Yep I just found a bug in the interrupt callback, fixed and sync'd. 

> Implementing hardware flow control should only need some configuration glue. The NXP drivers support it. I did not add it when creating machine_uart.c, because the respective pins were hardly available at the supported boards.

The imx side is not the problem it seems it's the module that needs to throttle the communications at higher baudrates, I tried to add hardware flow control, and it's easy to enable as you said, but the problem is specifying the CTS/RTS pins and configuring them, and yes I agree those tables really need to go.

> There is quite a bit at the MIMXRT port than could be reworked, especially the GPIO/Pin handling and

I couldn't agree more, question why aren't pins just simply defined like this ? Why the tables at all ?
```c
```
`mpconfigboard.h`
```c
#define MICROPY_HW_UART1_TX             (IOMUXC_GPIO_AD_B0_12_LPUART1_TX)
#define MICROPY_HW_UART1_RX             (IOMUXC_GPIO_AD_B0_13_LPUART1_RX)

#define MICROPY_HW_SPI1_SCK             (IOMUXC_GPIO_SD_B0_00_LPSPI1_SCK)
#define MICROPY_HW_SPI1_MOSI            (IOMUXC_GPIO_SD_B0_03_LPSPI1_SDI)
```

BTW do you happen to know if there's any reason why using LPUART3 on those specific pins wouldn't work or would require configuring GPR2x or something like that ? So with the exact same code, bluetooth doesn't work on another board, the only difference is the pinout.

```c
    { IOMUXC_GPIO_B0_08_LPUART3_TX }, { IOMUXC_GPIO_B0_09_LPUART3_RX }, \
```

iabdalkader commented:
> So with the exact same code, bluetooth doesn't work on another board, the only difference is the pinout.

@robert-hh I found the problem, it's actually the BT regulator on/off pin, it seems there's a problem in `machine_pin_set_mode`, for some reason it can't configure the pin as GPIO out. I see that it's missing the pad/mux config, that might it. FWIW the pin is `GPIO_AD_B0_08`.

robert-hh commented:
> for some reason it can't configure the pin as GPIO out.

There was a PR merged recently to fix that. If you rebase to master, it should work.

robert-hh commented:
>  Why the tables at all ?

To enable configuration of the device pins based on the device number. e.g. UART(1) allows to point into the table for the RX and TX GPIO. In the MIMXRT port, the device pins are hardwired.

iabdalkader commented:
> To enable configuration of the device pins based on the device number. e.g. UART(1) allows to point into the table for the RX and TX GPIO. In the MIMXRT port, the device pins are hardwired.

The pins can be used directly (see `stm32/uart.c` for example) or better that table can be statically allocated in `machine_uart.c`, for example see `pyb_i2c_obj` in `stm32/pyb_i2c.c`, much better cleaner, and will remove dynamic allocation of uart objects, and all the tables from every boardconfig file.

robert-hh commented:
> The pins can be used directly (see stm32/uart.c for example) or better that table can be statically allocated in machine_uart.c,  .....

 I know these, and they do not look better for me, only different. Not to mention that the tables have to different for every MCU.

iabdalkader commented:
> I know these, and they do not look better for me, only different. Not to mention that the tables have to different for every MCU.

I think they are better because:
* You only need to define the used pins in `mpconfigboard.h` and not the whole, mostly empty, table. This will massively simplify the `mpconfigboard.h` files, as there will only be 1 table in each `machine_*.c` file.
* You can reference the peripheral by it's actual ID vs the map from logical -> physical (which is very confusing btw).
* Static table will avoid the dynamic allocation of the peripheral (GC not being involved is always good).

I can probably think of more reasons, but simplifying and cleaning up config file I think is enough.

robert-hh commented:
- There is actually only one table in each machine_*.c file. The content is defined in the mpconfigboard.h file.
- The ordering of the table in mpconfigboard.h could be by logical ID, removing the empty slots.
- There is no problem with dynamic allocation. It's Python. And e.g. the UART buffers have anyhow to be allocated dynamically.

But that's not the change I'm thinking of. I would prefer to use the information which is in the pin-af data structures, such that there is no need to tediously collect the pin information for the mpconfigboard.h. making the *_af.csv files was quite a work. So it's better to derive the GPIO name related to a signal from that table, in one or the other way.

iabdalkader commented:
> But that's not the change I'm thinking of. I would prefer to use the information which is in the pin-af data structures, such that there is no need to tediously collect the pin information for the mpconfigboard.h

@robert-hh There's no need to collect anything, the information is already there in every `mpconfigboard.h`, you just need to move them from tables to single pin defines, and then add the static tables to machine modules. This is how every other port works and I don't see anything wrong with this approach, in fact making this port look like other ports would have helped someone like me to quickly add support for something like CTS/RTS. I'm not sure what you mean by using the pin_af data structures, can you please start a new issue/discussion/RFC and maybe post your proposal and let's see some examples ?

@ dpgeorge Pending issues:

- The `cywbt_wait_cts_low` function in `cywbt.c` should not change IO config, because the pin is still readable.
- In `cywbt.c` every `mp_hal_pin` should be replaced with `cyw43_hal_pin`

iabdalkader commented:
rebased.

iabdalkader commented:
@dpgeorge Maybe this one can be merged too ? Possibly after the next release, but it's not as invasive as it seems.

dpgeorge commented on drivers/cyw43/cywbt.c:
```diff
@@ -29,42 +29,48 @@
+
 STATIC void cywbt_wait_cts_low(void) {
-    mp_hal_pin_config(pyb_pin_BT_CTS, MP_HAL_PIN_MODE_INPUT, MP_HAL_PIN_PULL_UP, 0);
+    mp_hal_pin_config(CYW43_PIN_BT_CTS, MP_HAL_PIN_MODE_INPUT, MP_HAL_PIN_PULL_UP, MP_HAL_PIN_AF_GPIO);
```
Passing 0 in the last argument is simply because it's an unused argument.  There is no such thing as a GPIO alt function.

So, either keep it as 0, or have a constant named `MP_HAL_PIN_AF_NONE`.

iabdalkader replied:
The `alt` argument is used in `mp_hal_pin_config` in `stm32` port, and I defined `MP_HAL_PIN_AF_GPIO` as 0, for `mimxrt` the GPIO alt function is `5`

dpgeorge replied:
But the alt argument is only needed on stm32 if MP_HAL_PIN_MODE_ALT is used.  And it should be this way on other ports as well.  If mimxrt needs to set alt to 5 then it should do that internally in the function when MP_HAL_PIN_MODE_INPUT/OUTPUT is passed in as the mode.

Really the alt argument should be removed altogether from this function, and functions like mp_hal_pin_config_alt used for non-GPIO cases.

iabdalkader replied:
> . If mimxrt needs to set alt to 5 then it should do that internally in the function

This is possible, but I was trying to avoid it by just defining the GPIO AF value to 0 or 5 (for mimxrt), but I will revert this to `0` and handle it in mimxrt. Note switching to GPIO to read the pin is not actually needed, at least for the ports that currently support this driver.

iabdalkader replied:
I reverted `MP_HAL_PIN_AF_GPIO` to 0.

> Really the alt argument should be removed altogether from this function, and functions like mp_hal_pin_config_alt used for non-GPIO cases.

`mp_hal_pin_config_alt` only exists to support the `stm32` instance number (for example `AF_FN_UART, 1`), other ports just use `mp_hal_pin_config`. Anyway, `cywbt_wait_cts_low` is not portable, because it needs `AF_FN_UART`, I think it can be made portable if the pin mode is not changed, or if this function is made abstract and implemented by ports (for example `mp_bluetooth_hci_uart_poll_cts`).

dpgeorge commented on drivers/cyw43/cywbt.c:
```diff
@@ -29,42 +29,48 @@
 
 STATIC int cywbt_hci_cmd_raw(size_t len, uint8_t *buf) {
-    uart_tx_strn(&mp_bluetooth_hci_uart_obj, (void*)buf, len);
+    mp_bluetooth_hci_uart_write((void *)buf, len);
```
This is subtly different behaviour calling this function, because it now calls `mp_bluetooth_hci_controller_wakeup()`.  I'm not sure if that matters but maybe it's best to keep the original behaviour, somehow.

iabdalkader replied:
Short of adding a new higher-level function that only writes to UART with no calls to `wakeup()`, I'm not sure how to keep the original behaviour, but everything from that point forward calls  `mp_bluetooth_hci_uart_write` so it shouldn't be a problem, and FWIW I tested this change on `stm32` and `mimxrt` and everything seemed to continue to work normally.

iabdalkader replied:
One more thought about this, if `wakeup()` shouldn't be called before certain initialization is done, I think that should be handled/ensured by the lower-level functions (in the driver).

dpgeorge replied:
OK, I checked the code and I think it's fine to call wakeup here.  This `cywbt_hci_cmd_raw()` function is only called from within `mp_bluetooth_hci_controller_init()`, and for all cases that it's called the `pyb_pin_BT_DEV_WAKE` pin is low, which means the controller is awake.  That pin is driven high (to let the controller sleep) only at the end of that init function.

So, let's leave this as you have it.

dpgeorge commented on drivers/cyw43/cywbt.c:
```diff
@@ -29,42 +29,48 @@
             MICROPY_EVENT_POLL_HOOK
         }
-        buf[i] = uart_rx_char(&mp_bluetooth_hci_uart_obj);
+        buf[i] = mp_bluetooth_hci_uart_readchar();
```
Would be simpler to write this as:
```c
int c;
while ((c = mp_bluetooth_hci_uart_readchar()) == -1) {
    MICROPY_EVENT_POLL_HOOK
}
buf[i] = c;
```

dpgeorge replied:
But `mp_bluetooth_hci_uart_readchar()` calls `uart_rx_any()` before it tries to read (on stm32 at least).  So in the PR as it currently stands, any() is called at least twice for each character read.

The version I propose above should be equivalent to the original code in terms of the ordering and number of calls to `uart_rx_any()` and `uart_rx_char()`.  Unless I missed something?

iabdalkader replied:
> But `mp_bluetooth_hci_uart_readchar()` calls `uart_rx_any()` before it tries to read

Oh okay I see your point, changed.

dpgeorge commented on ports/mimxrt/mpbthciport.c:
```diff
@@ -0,0 +1,196 @@
+#if MICROPY_PY_BLUETOOTH
+
+#define debug_printf(...) // mp_printf(&mp_plat_print, "mpbthciport.c: " __VA_ARGS__)
+#define error_printf(...) mp_printf(&mp_plat_print, "mpbthciport.c: " __VA_ARGS__)
```
Please use `DEBUG_printf` and `ERROR_printf` to match elsewhere in the code base.

iabdalkader commented:
Are we any closer to resolving the pending conversations here ?

dpgeorge commented on drivers/cyw43/cywbt.c:
```diff
@@ -81,10 +88,10 @@ STATIC int cywbt_hci_cmd_raw(size_t len, uint8_t *buf) {
             MICROPY_EVENT_POLL_HOOK
         }
-        buf[i] = uart_rx_char(&mp_bluetooth_hci_uart_obj);
+        buf[i] = mp_bluetooth_hci_uart_readchar();
```
Please also change this to the same pattern as above.

iabdalkader replied:
Can we instead save a line and:

```c
    for (int c, i = 0; i < sz; ++i) {
        while ((c = mp_bluetooth_hci_uart_readchar()) == -1) {
            MICROPY_EVENT_POLL_HOOK
        }
        buf[i] = c;
    }
```
?

dpgeorge commented on ports/stm32/mpbthciport.c:
```diff
@@ -210,6 +210,11 @@ int mp_bluetooth_hci_uart_set_baudrate(uint32_t baudrate) {
     return 0;
 }
 
+int mp_bluetooth_hci_uart_any(void) {
```
I think this function can be deleted now.

dpgeorge commented:
I tested this on PYBD-SF2/SF6 with the BLE multitests, and it works.

dpgeorge commented:
> Pending issues:
> 
>     * The `cywbt_wait_cts_low` function in `cywbt.c` should not change IO config, because the pin is still readable.
> 
>     * In `cywbt.c` every `mp_hal_pin` should be replaced with `cyw43_hal_pin`

@iabdalkader I agree these points can be improved (but they were not necessary to get this PR merged).  Feel free to have a go at fixing them, otherwise I'll eventually get to them.

iabdalkader commented:
Well at least now the driver is usable by other ports so it should be fine for now, but may get around to it. For the first point, adding `mp_bluetooth_hci_uart_poll_cts` might be the safest option.

Are there any plans (or reasons not) to move `cywbt.c` to `lib/cyw43-driver/` ?

dpgeorge commented:
> Are there any plans (or reasons not) to move `cywbt.c` to `lib/cyw43-driver/` ?

We can do that.  It changes the licensing but I don't think there's really a problem with that.  Would be neater to keep everything in lib/cyw43-driver.  (Also rp2 uses a completely different BT backend, but again that doesn't impact moving this file there.)

iabdalkader commented:
> > Are there any plans (or reasons not) to move `cywbt.c` to `lib/cyw43-driver/` ?
> 
> We can do that. It changes the licensing but I don't think there's really a problem with that. Would be neater to keep everything in lib/cyw43-driver. (Also rp2 uses a completely different BT backend, but again that doesn't impact moving this file there.)

Following-up on this, I moved `cywbt.c` to `cyw43-driver` in PR  https://github.com/georgerobotics/cyw43-driver/pull/123. The PR also makes the firmware address and size configurable. This allows us to move the bt firmware blob to QSPI for example (if needed) to save flash space.
